Custom url navigations

vanillaHafer opened the pull request:
### Context

Currently, Spina::NavigationItems can only link to other main Spina pages and they cannot be edited. It would be nice if Custom URLs were able to be added in, to allow more customization to the user.

### Changes proposed in this pull request

The 2 main changes in this PR are:

1. Being able to edit existing navigation items
<img width="799" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/39d116ca-a3d1-4296-8aef-da42d7ba9fd6">
2. Being able to use custom URLs with their labels, instead of having to select from the list of pages
<img width="695" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/8dae8957-d7c2-43ee-ad72-bd982c277fee">

### Guidance to review

Everything existing functions as normal. All changes are simply additions.

Bramjetten commented on app/models/spina/navigation_item.rb:
```diff
@@ -25,6 +25,14 @@ def menu_title
       end
     end
 
+    def materialized_path
```
This could be rewritten as:
```ruby
def materialized_path
  page&.materialized_path || url
end
```

or like this if you prefer something a bit more explicit:

```ruby
def materialized_path
  return page.materialized_path if page
  url
end
```

vanillaHafer replied:
Great suggestion. I went ahead and made those corrections.

Bramjetten commented on app/models/spina/navigation_item.rb:
```diff
@@ -11,8 +11,40 @@ class NavigationItem < ApplicationRecord
+    validate :url_or_page_presence
 
-    delegate :menu_title, :materialized_path, :draft?, :homepage?, to: :page
+    delegate :draft?, :homepage?, to: :page
```
I think these also need their own method now, defaulting to false when there's no page. We might be able to get away with `allow_nil: true` here.

vanillaHafer replied:
Good call! That was an oversight on my part. Went ahead and added the `allow_nil: true` to take care of this.

Bramjetten commented on test/dummy/db/schema.rb:
```diff
@@ -124,12 +124,14 @@
     t.datetime "created_at", precision: nil
     t.datetime "updated_at", precision: nil
+    t.string "url"
+    t.string "url_label"
```
Why not `menu_title` like for pages? How do you handle translations?

mattpolito replied:
We ended up making changes so there is a `url_title` attribute that get translated through mobility.

Bramjetten reviewed (commented):
Thanks for working on this! I like the idea here, but I'm not sure about the UI implementation yet. Few things to consider:
- How do you handle translations? This goes for the label, but also the URL
- Instead of having a button "Add menu item", perhaps it should be two buttons: "Add page" and "Add URL" which will then have their own forms.

mattpolito commented:
> * Instead of having a button "Add menu item", perhaps it should be two buttons: "Add page" and "Add URL" which will then have their own forms.

@Bramjetten I guess it depends on if we have that design language elsewhere in the site. Personally I would think that I'm going to this section to add a navigation item... so already the language of "Add menu item" differs but at least it's a singular thing. If we add choices here, I would think that someone would always choose "Add URL" as it makes the most sense for a nav item. 

For confusion sake, I would suggest that there be a single link to open a form where you'd then choose the kind of thing you want to create. In the end you're always creating "A menu item".

mattpolito commented:
@Bramjetten Also we noticed that a test failure was introduced by a previous commit so you'll see the fix for that in here as well.

Bramjetten commented:
@mattpolito I've noticed some flakiness with a certain Resource test, so I might have to fix that anyway.

mattpolito commented:
@Bramjetten do you have any additional feedback on this one?

Bramjetten commented:
I would very much like to have this feature, so let's work on getting it in Spina!

Couple of things I notice in the UI in no particular order:
- The "Choose page" select is a simple dropdown which is hard to use when you have many pages. Maybe we can borrow PageLink's select with search functionality.
- Using a toggle to switch forms feels odd. Also, the window jumps when you do that.
- The input fields don't have autofocus
- Perhaps instead of a toggle, it can simply be a link to a different form
- There's no validation on adding a URL
- It doesn't support translations
- If you can change the label for custom URLs, shouldn't you be able to do that for pages as well?
- I feel the navigation items themselves should have a little indicator to tell you it's a custom URL and not a page

Couple of things I notice in the code:
- Why not use/update the SwitchComponent?
- Can we have one migration instead of two?
- Instead of altering `navigation_item_params` in the controller, I'd prefer to use Turbo Frames to change the form instead of showing/hiding input fields with JS.

mattpolito commented:
> * The "Choose page" select is a simple dropdown which is hard to use when you have many pages. Maybe we can borrow PageLink's select with search functionality.

I completely agree with this needing improvement. However it is not part of adding a custom url so I'd leave it for another PR.

> * Using a toggle to switch forms feels odd. Also, the window jumps when you do that.

Clearly you haven't been a fan of the toggle and I don't think a separate button is the right choice either. How do you feel about that model opening with tabs? One for "Page" and another for "URL"?

> * The input fields don't have autofocus

This can be easily addressed

> * Perhaps instead of a toggle, it can simply be a link to a different form

See above comment regarding toggle/form

> * There's no validation on adding a URL

Technically there is a validation. At least one that says it has to be populated if you select the custom url option. Are you referring to something more specific validating the URL itself? Also due to how modals are implemented currently, It does not allow for errors to come back from turbo. We've been wanting to address this anyway. Again, I think that'd be change but not part of this PR.

> * It doesn't support translations

We added translations the other day. Did you not see that change or are you referring to something else?

> * If you can change the label for custom URLs, shouldn't you be able to do that for pages as well?

We didn't make that change because currently you have the mechanism to change a page label on the page itself. Not sure if we want to immediately tackle this but we think it should actually live here in the navigation item. That way we get a uniform experience as well as the ability to have have pages in multiple navigations that can have a different name than the singular one defined on the page.

> * I feel the navigation items themselves should have a little indicator to tell you it's a custom URL and not a page

I agree with this.

> * Why not use/update the SwitchComponent?

If you don't like the toggle, then there's no reason to use the switch component. So we just need to make a decision on UX.

> * Can we have one migration instead of two?

We *CAN* however we're addressing two different tables so it seems appropriate to have two different migrations. With that said... Is there a reason we're using integer migration versions vs timestamp?

> * Instead of altering `navigation_item_params` in the controller, I'd prefer to use Turbo Frames to change the form instead of showing/hiding input fields with JS.

This again refers to the decision that needs to be made about the UX.

Overall... we can get the changes in that we all agree on and let's discuss how you feel about the form UX so we can move forward with the other changes.

vanillaHafer commented:
We made a few changes recently based on your feedback. We got away from the toggle mechanism and have two separate forms using tabs:
<img width="1439" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/7f856489-51a5-48c6-9d4f-3a85fc47d1b7">
<img width="1440" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/44567b91-9518-4a56-b35e-21740b2e5b00">

We also added in the indicator showing if a navigation item is an external link:
<img width="1440" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/e0c49602-c2a2-4aa7-94e8-a5dfa19ae36c">

We also addressed validating the url and and url_title in the form:
<img width="1440" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/5f367cf1-4522-4e74-9012-051714bce173">

We also implemented autofocus for the url form.

Whenever you get a chance, feel free to look over the changes and let us know if there's any other feedback you have!

Bramjetten commented:
Very cool, nice!

I think we can improve the UI a bit more. I was thinking about something like this:
<img width="820" alt="CleanShot 2023-07-31 at 16 27 46@2x" src="https://github.com/SpinaCMS/Spina/assets/423116/a6d66e75-9d00-48ee-a478-59fd0663ef55">

This way, adding a page to the menu is still the primary way of adding items while simultaneously making it very clear what the other option is meant for.

As for the indicator on the item itself, perhaps we can use this icon with a small tooltip: 
<img width="34" alt="CleanShot 2023-07-31 at 16 30 20@2x" src="https://github.com/SpinaCMS/Spina/assets/423116/63733d4c-8886-4d66-9479-80206df860a7">

Excuse me if I missed it, but have you taken a look at adding something for translations?

vanillaHafer commented:
I went ahead and implemented those UI changes, so hopefully they are looking more like what you were expecting.

### Removed tabs in favor of links
<img width="1437" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/cd3735e7-9c7e-4eed-99ef-03a939011869">
<img width="1440" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/a5586031-3985-4b6b-b5db-06b650ac2090">

### Changed indicator
<img width="1122" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/72907b33-2c67-4d7e-be92-e896fc8b8073">

### Translations

As far as translations go, we're handling translations for the UI elements the same as all the other elements, by adding to the `.yml` files under `config/locales/`. And then regarding actual URL Title translations we are using a translation table for navigation items. As far as we understand, that's enough to handle translations, but please let us know if we are missing something.

Bramjetten commented:
This is great! 

What I mean with the translations is that I don't think the UI currently supports editing multiple translations. How can end-users add multiple translations for a menu item?

mattpolito commented:
> How can end-users add multiple translations for a menu item?

How is that working anywhere else? I don't recall seeing that functionality.

Bramjetten commented:
> How is that working anywhere else? I don't recall seeing that functionality.

<img width="273" alt="CleanShot 2023-07-31 at 21 23 05@2x" src="https://github.com/SpinaCMS/Spina/assets/423116/8fbbf0d9-d449-42d7-9bce-9480cbd70088">

vanillaHafer commented:
Since we don't use translations for our application, the translation feature was a bit foreign to us. Thanks for being patient with us as we learn more about Spina to get this feature rolling.

Looking at the current implementation I was able to throw something together for the Navigations/Navigation Items.

As you can see on the Navigations screen we now have a Translation Component in the upper right:
<img width="1552" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/37673690-0a35-4bcf-8ba2-55a74abfc82b">

Which will display Navigation Items using their correct translations:
<img width="1552" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/be24e2e4-07b7-42a8-abae-7a57c2482398">

When adding in a navigation, we can choose which translation we are adding:
<img width="530" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/10bef287-ff59-4b09-a1cb-c432e57bb64e">

After adding in a navigation item and giving it, say, an English translation first, we can now see that some translations are missing url_titles:
<img width="1552" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/3c359e11-05b0-4b57-977e-803c3668c2a3">

Which is also seen when editing that particular navigation item:
<img width="528" alt="image" src="https://github.com/SpinaCMS/Spina/assets/27235783/8fb0b509-c0df-43da-88a0-1c2f1212d15a">

As usual, please give us any feedback you have on this so we can correct anything you think should change!

Bramjetten commented:
I think we can improve the UI for switching locales. I think it’s best if we make it possible to fill in all translations in one go, as that’s probably the most common workflow. Perhaps for now it’s best to simply list the different locale fields from top to bottom and improve the UI further in a different PR. That way we can get this merged now and enable it for everyone :)

mattpolito commented:
>  I think it’s best if we make it possible to fill in all translations in one go, as that’s probably the most common workflow.

@Bramjetten is there an example of this workflow currently in the codebase?

Bramjetten commented:
No there isn’t, but the easiest way is to simply loop over all locales in Spina.config.locales and render the fields manually.

mattpolito commented:
Since that is a new UI concept, I would really suggest it be separate from this PR. The data flow wouldn't change so doing it at a later time would not make a difference.

Bramjetten commented:
@mattpolito Agreed that we can merge this into the main branch now and clean up the UI in a separate PR. I’ll create an issue for it.

Bramjetten commented:
@mattpolito Do you know why the tests are failing?

mattpolito commented:
@Bramjetten thats an interesting failure. I'll look at it or have @vanillaHafer double check it. It's failing on Minitest so it may be a CI issue.

mattpolito commented:
@Bramjetten suite on this branch was passing. The error that is currently happening seems to have been introduced from #1273
